### ISSUE TITLE
Support for VAB version 7

### ISSUE THREAD

**Issue by Nisto** [vgmtrans/vgmtrans#243]
**Is your feature request related to a problem? Please describe.**
One minor but crucial difference between the common versions 5 and 7 of the PS1 VAB format appears to be the change from offsets to sizes in the VAG table. Currently, it seems VGMTrans assumes offsets (as in VAB format version 5). In the log window, VGMTrans shows e.g. "[Vab] VAG #4 pointer (offset=0x000250B0, size=75792) is invalid."

**Describe the solution you'd like**
For version 5, parse the VAG table values as offsets; for version 7, parse the values as sizes.

**Additional context**
None.

**Comment by loveemu** [comment]
I'm just curious, can you name a specific game title that uses VAB version 5 and VAB version 7?　If and when this request is fulfilled, it would be helpful to test the functionality.

**Comment by Nisto** [comment]
Well, this is embarrassing. It's been a little while since I opened this issue, and I honestly can't remember where I was supposed to have found version 5 VAB files with offsets in the VAG table... In fact, when looking today, I only managed to find version 5 VAB files in the PS1 SDK, and those all contain sizes in the VAG table. Even some of the earliest games don't seem to have any version 5 VAB files.

However, I am curious why VGMTrans seems to parse the VAG table as offsets, because after double checking today, it seems all VAB files, regardless of version, store sizes rather than offsets. I don't know, maybe I've misinterpreted the VGMTrans code, and the VAG pointer error quoted in my original post is not necessarily a cause of how the VAG table values are interpreted? Maybe a truncated file or something.. ? Again, I have no idea what I was trying to load back when I opened this issue, so I can't double check I'm afraid.

For posterity's sake, here's what I was able to find:

Games with VAB version 6:
* Biohazard
* Crash Bandicoot
* Crime Crackers
* Cyber Sled
* King's Field
* Ridge Racer
* Tomb Raider

Games with VAB version 7:
* Alundra
* Biohazard 2
* Digimon World
* Echo Night
* Genso Suikoden II
* Metal Slug - Super Vehicle-001
* Silent Hill

**Comment by Meerkov** [comment]
I'm currently looking at a game 3Xtreme, which has version 7 files.

E.g. one file called FATHEAD.SND. This file doesn't appear to parse correctly. Is there any reference material on the file formats here? I can maybe try to look at this with some additional info.

**Comment by Nisto** [comment]
The following citations (emphasis added by me) are from the official PlayStation Technical Reference version 4.1, released in December 1997. I can confirm the third point is valid for VAB versions 5, 6 and 7, at least based on VAB files that are included across various versions of the official SDKs. However, note that the first entry in a VAG table is usually 0 - there is no mention as to exactly why that is.

"File Formats", chapter 4 "Sound - VAB: PS Sound Source Data":
* each VAB file can contain up to 254 VAG files
* VAG (N) byte count: any (up to 516,096)
* VAG offset table contains 3-bit right-shifted VAG data ***size*** stored in short (16 bit). For example:

| VAG# | 0 | 1 | 2 | ... |
|----------|:-------------|:------|:------|:------|
| VAG offset table | 0x1000 | 0x0800 | 0x0200 | ... |
| Actual size | 0x8000 | 0x4000 | 0x1000 | ... |
| Offset | 0x8000 | 0xc000 | 0xd000 | ... |

**Comment by Meerkov** [comment]
https://github.com/vgmtrans/vgmtrans/pull/283
Things look a little funky, but I think I'm starting to grasp it. Please take a look and make sure I'm not going off in the wrong direction.

**Comment by Meerkov** [comment]
One possible reason you may be seeing instruments of size "0" as the first instrument, is that at least from what I've read on other bugs, an empty instrument is considered a valid instrument. 

(IMO: Random hypothesis on why a developer might include a size 0 instrument as the first instrument: since these files can be changed dynamically when sent to the SPU, it might be a design decision to allow silencing instruments dynamically by swapping the instrument used to 0, rather than turning the volume to 0. There might be efficiency reasons that processing an empty instrument is faster than processing a normal instrument and throwing the results away)

**Comment by Meerkov** [comment]
The information that I can find, and from reading the code, is that "Offset" is simply just a misnomer. The documentation calls these offsets, but they are clearly described as sizes, and used as sizes. Same thing in the VGMTrans code. They are treated as sizes.

I'll keep my eyes open for version 5 VAB, but everything I've read suggests that only version 6 and 7 are likely in use in practice....

Maybe the best result would be to add in an error-catcher for parsing VAB version 5 so that someone eventually complains about it, and we can find out if the format is actually different or not.

**Comment by loveemu** [comment]
I agree with Meerkov's view. "Offset" is a misnomer and was recently corrected in PR #283.

> In fact, when looking today, I only managed to find version 5 VAB files in the PS1 SDK, and those all contain sizes in the VAG table. Even some of the earliest games don't seem to have any version 5 VAB files.

Does this mean that there are no specific cases where we need to add support for now?

**Comment by Nisto** [comment]
I am in complete agreement with you guys. "VAG offset table" is obviously a misnomer. Even though they refer to the table exactly as "VAG offset table" every time (4 times that I can see) in that technical reference document, what actually stands out to me is the part that actually defines **what it contains** ("VAG offset table contains 3-bit right-shifted VAG data size stored in short"). The "name" of the table is open to interpretation I guess, which is never good..

<br>

@Meerkov 
> The information that I can find, and from reading the code, is that "Offset" is simply just a misnomer. The documentation calls these offsets, but they are clearly described as sizes, and used as sizes. Same thing in the VGMTrans code. They are treated as sizes.

Thanks for double checking this. I could've sworn it looked like VGMTrans parsed the table values as offsets, but misnamed variables probably just mislead me, then. I'd probably try to update those names though, yikes!

<br>

@Meerkov 
> I'll keep my eyes open for version 5 VAB, but everything I've read suggests that only version 6 and 7 are likely in use in practice....

FWIW, I think I've figured out how I ended up with those VAB "version 5" files back when I opened this issue. One of my conversion scripts for Biohazard 3 was converting banks incorrectly. It has a bank format very closely based on VAB, so the script simply attempts to restore the files into standard VAB format. The most notable difference is, the tables *actually* contain *offsets* -- which my script wasn't converting to sizes. Also, there is no format version specified in the Biohazard 3 format, and my script was probably based on VAB v5 files from the official SDKs, used to fill in missing values. My sincere apologies for the confusion!

**Comment by mikelow** [comment]
This appears to be resolved.
